### ISSUE TITLE
Size of Torch .so libraries with generics

### ISSUE THREAD

**Issue by borisfom** [torch/torch7#841]
@soumith:
I have a report that cutorch.so has major issues loading on Android due to its increased size with the introduction of double math (and half, in my case: 450M, used to be 90M). I suggest implementing a switch that would allow to only build particular types. I have used TH_GENERIC_USE_HALF in my PR for HalfTensor - should we add (TH_GENERIC_USE_DOUBLE, TH_GENERIC_USE_SHORT, TH_GENERIC_USE_BYTE) for types that may not be necessary, throughout Torch modules (torch7->cutorch->cunn) ?

**Comment by borisfom** [comment]
I am leaning towards TH_GENERIC_NO_XXX macros. What would be the most sensible default set? Currently we have char and unsigned char, short, int and long in integers - looks like a huge overkill as each of them means extra copy of all Torch generics, and combinatorial increase of copy methods.
'double' probably should be in the default set though, but I plan to try turning it off for Android.

**Comment by soumith** [comment]
the most sensible default set is to have all of them enabled.
For android or a custom install, defining:

THC_GENERIC_NO_* should be good.

The minimum working set is Byte, Float and Long if i remember.

**Comment by borisfom** [comment]
Thanks Soumith! Yes changing default would definitely not be a good idea - minimum set is what I was in fact looking for to provide a group macro.

**Comment by borisfom** [comment]
@soumith: actually, while implementing those switches i figured a smarter way to go around type expansion control would probably be from CMakeLists.txt, which commands macros like TH_GENERIC_USE_XX : 
SET(TH_GENERIC_DEFAULT_TYPES Float Double Char Byte Short Int Long)
SET(TH_GENERIC_MINIMUM_TYPES Float Byte Long)

IF (ANDROID)
  SET(TH_GENERIC_TYPES TH_GENERIC_MINIMUM_TYPES)
ELSE()
  SET(TH_GENERIC_TYPES TH_GENERIC_DEFAULT_TYPES)
ENDIF()

foreach(TH_TYPE ${TH_GENERIC_TYPES})
  LIST(APPEND GENERIC_FLAGS "-DTH_GENERIC_USE_${TH_TYPE}")
endforeach()
# use GENERIC_FLAGS later to append to both C and C++ flags

**Comment by soumith** [comment]
i dont want that logic to be dependent on cmake really, that is quite bad, considering that the THC headers are used by THCUNN as well.

The only file in the whole PR I want you to change is https://github.com/torch/cutorch/blob/master/lib/THC/THCGenerateAllTypes.h

**Comment by soumith** [comment]
and https://github.com/torch/cutorch/blob/master/lib/THC/THCGenerateFloatTypes.h

**Comment by borisfom** [comment]
Right, let's keep it simple : https://github.com/torch/torch7/pull/842
It's a bit more files than that to ensure it actually compiles when one of those macros are defined, but I think it does what you meant.
THC to follow.
